### DISCUSSION TITLE
an error when i use http package in endpoint

### DISCUSSION THREAD

**Discussion by Metaildust** [serverpod/serverpod#2281] [Q&A] [answered]
I Want to use a third-party OSS service. But when I tried to obtain the STS token for a third-party OSS service in the endpoint, my http request received a response. body, which should theoretically be of type string. But when I tried to send this string to the front-end, an error occurred, indicating that the type _Map<String, dynamic> is not an sub of List(click in you can see like _Map＜String, dyamic＞ as List). But if I use JsonDecode directly on the backend to load response.body as a map type, and then there is a deserialization error. The same code, if I run it in Flutter, everything works fine, but if I run it in Endpoint, this error will occur. Is this an type error between flutter and endpoint?

flutter: Failed call: sts.getOssSts
flutter: FormatException: No deserialization found for type dynamic
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: FormatException: No deserialization found for type dynamic
#0      SerializationManager.deserialize (package:serverpod_serialization/src/serialization.dart:86:5)
#1      Protocol.deserialize (package:for_one_gen_client/src/protocol/protocol.dart:106:18)
#2      Protocol.deserialize.<anonymous closure> (package:for_one_gen_client/src/protocol/protocol.dart:101:44)
#3      MapBase.map (dart:collection/maps.dart:82:28)
#4      Protocol.deserialize (package:for_one_gen_client/src/protocol/protocol.dart:100:28)
#5      SerializationManager.decode (package:serverpod_serialization/src/serialization.dart:40:12)
#6      parseData (package:serverpod_client/src/serverpod_client_shared_private.dart:21:31)
#7      ServerpodClient.callServerEndpoint (package:serverpod_client/src/serverpod_client_io.dart:76:16)
<asynchronous suspension>
#8      StsController.refreshStsToken (package:for_one_gen_flutter/app/homepage/signed_homepage.dart:21:35)
<asynchronous suspension>

Code:
Future<Map＜String,dynamic＞> getOssSts(Session session) async {
    String generateNonce() {
      final random = Random.secure();
      final bytes = List<int>.generate(16, (_) => random.nextInt(256));
      return base64.encode(bytes);
    }

    String formatTimestamp(DateTime timestamp) {
      return '${timestamp.toUtc().toString().substring(0, 19).replaceAll(' ', 'T')}Z';
    }

    String getSignature(String accessKeySecret, String stringToSign) {
      final key = utf8.encode(accessKeySecret);
      final hmac = Hmac(sha256, key);
      final signature = hmac.convert(utf8.encode(stringToSign));

      return signature.bytes
          .map((byte) => byte.toRadixString(16).padLeft(2, '0'))
          .join();
    }

    Future<Map(String,dynamic)> assumeRole(
        String accessKeyId,
        String accessKeySecret,
        String roleArn,
        String roleSessionName,
        int durationSeconds,
        String policy) async {
      const endpoint = "https://sts.cn-chengdu.aliyuncs.com";
      const apiVersion = "2015-04-01";
      const action = "AssumeRole";

      final now = DateTime.now().toUtc();
      final formattedTime = formatTimestamp(now);
      final signatureNonce = generateNonce();

      const canonicalQueryString = "";
      final requestPayload = json.encode({
        "RoleArn": roleArn,
        "RoleSessionName": roleSessionName,
        "DurationSeconds": durationSeconds,
        "Policy": policy,
      });

      final hashedRequestPayload =
          sha256.convert(utf8.encode(requestPayload)).toString();
      final canonicalHeaders =
          "host:${endpoint.replaceAll(RegExp(r'https?://'), '')}\nx-acs-action:$action\nx-acs-content-sha256:$hashedRequestPayload\nx-acs-date:$formattedTime\nx-acs-signature-nonce:$signatureNonce\nx-acs-version:$apiVersion\n";
      const signedHeaders =
          "host;x-acs-action;x-acs-content-sha256;x-acs-date;x-acs-signature-nonce;x-acs-version";
      final canonicalRequest = """
POST
/
$canonicalQueryString
$canonicalHeaders
$signedHeaders
$hashedRequestPayload""";
      final hashedCanonicalRequestBytes =
          sha256.convert(utf8.encode(canonicalRequest)).bytes;
      final hashedCanonicalRequest = hashedCanonicalRequestBytes
          .map((byte) => byte.toRadixString(16).padLeft(2, '0'))
          .join();
      final stringToSign = "ACS3-HMAC-SHA256\n$hashedCanonicalRequest";
      final signature = getSignature(accessKeySecret, stringToSign);

      final authorization =
          "ACS3-HMAC-SHA256 Credential=$accessKeyId,SignedHeaders=$signedHeaders,Signature=$signature";

      final response = await http.post(
        Uri.parse(endpoint),
        body: requestPayload,
        headers: {
          "Authorization": authorization,
          "x-acs-action": action,
          "x-acs-content-sha256": hashedRequestPayload,
          "x-acs-date": formattedTime,
          "x-acs-signature-nonce": signatureNonce,
          "x-acs-version": apiVersion,
          "host": endpoint.replaceAll(RegExp(r'https?://'), ''),
          "Content-Type": "application/json",
        },
      );

      if (response.statusCode == 200) {
        
        return jsonDecode(response.body);
      } else {
        throw MyException(
            message:
                'Failed to assume role. Status code: ${response.statusCode}, Body: ${response.body}');
      }
    }

    var userid = await session.auth.authenticatedUserId;
    String policy =
        '{"Version":"1","Statement":[{"Effect":"Allow","Action":["oss:GetObject","oss:PutObject","oss:DeleteObject"],"Resource":["acs:oss:*:*:metaildust-1/user-data/user$userid/*"],"Condition":{"StringEquals":{"oss:Prefix":"user-data/"}}}]}';

    Map＜String,dynamic＞ result = await assumeRole(
        'LTAI5tQpq6**************',
        'a4cCR*************************',
        'acs:ram::1938812721441147:role/ramosstest',
        'user${userid!}',
        3600,
        policy);
    return result;
  }

What should be response.body like:
'' {
  "RequestId": "6894B13B-6D71-4EF5-88FA-F32781734A7F",
  "AssumedRoleUser": {
    "AssumedRoleId": "34458433936495****:alice",
    "Arn": "acs:ram::123456789012****:role/adminrole/alice"
  },
  "Credentials": {
    "SecurityToken": "********",
    "Expiration": "2015-04-09T11:52:19Z",
    "AccessKeySecret": "wyLTSmsyPGP1ohvvw8xYgB29dlGI8KMiH2pK****",
    "AccessKeyId": "STS.L4aBSCSJVMuKg5U1****"
  }
}''

Maybe i used unsupported return type, but if i just directly return the response.body to the flutter as String. There is still an ever

**Comment by SandPod** [comment, marked as answer]
Hey @Metaildust 👋 

The endpoint serialization has a limited type support. You can see what types are supported here: https://docs.serverpod.dev/concepts/working-with-endpoints#passing-parameters.

All types used in a `Map` need to be strictly typed, so a `dynamic` value, such as you are defining as return value above, is not supported.

You will probably need to deserialize the response on the server side before passing it to the client to get away from the `dynamic` value. Passing it can either be done through a Serializable model, which i would recommend, or some other type that you prefer.

**Reply by Metaildust** [reply]
But will it work if i pass 『response.body』 as String? I tried but it didn't work. If it's feasible to do it this way I will try again and post error to see if i did somethong wrong

**Reply by SandPod** [reply]
That should work, just ensure that you update the return type of your endpoint to a String as well.
